feat: form phase 2

chohongm opened the pull request:
### Ticket
https://sendbird.atlassian.net/browse/AC-3349

### Changelog
- Improved user interface and experience of form message feature

### Note
discussion: https://sendbird.slack.com/archives/C0772Q0BB8Q/p1721093737987559?thread_ts=1720573872.559059&cid=C0772Q0BB8Q

Also belows are resolved in this PR:
- https://sendbird.atlassian.net/browse/AC-3111
- https://sendbird.atlassian.net/browse/AC-3010
- https://sendbird.atlassian.net/browse/AC-2782

chohongm committed:
apply design request

chohongm committed:
format

chohongm committed:
format

chohongm committed:
Align text in widget message input

chohongm committed:
fix margin for typing dots

chohongm commented on src/components/CustomTypingIndicatorBubble.tsx:
```diff
@@ -67,6 +67,7 @@ const TypingDot = styled.span`
 const Root = styled.div`
   display: flex;
   align-items: flex-end;
+  margin-top: 16px;
```
fixes https://sendbird.atlassian.net/browse/AC-3010

chohongm commented on src/components/CustomChannelComponent.tsx:
```diff
@@ -111,7 +111,8 @@ const Root = styled.div<RootStyleProps>`
     .sendbird-message-input--placeholder {
-      top: 9px;
+      top: 8px; // because top padding is 8px and the input has no border.
+      line-height: 24px; // because top & bottom padding is 8px each and the input height is 40px.
```
fixes https://sendbird.atlassian.net/browse/AC-3111

bang9 replied:
이거 absolute 가운데 정렬 하려는거면, 아래 스타일 주는게 낫지 않을까요? 눈대중으로 픽셀 맞추는건 웬만하면 없으면 좋을 것 같습니다.
```css
top: 50%;
transform: translateY(-50%);
```

chohongm replied:
오 absolute 도 % 로 되는군요! 
      top: 0;
      transform: translateY(50%);
로 수정했습니다, 감사합니다!

bang9 replied:
`top: 0 + translateY: 50%` 는 일반적인 css 계산상으로 가운데 정렬이 안되지 않나요?

bang9 replied:
top 50% + translateY 50% 는
요소의 top 위치를 부모의 50% 위치로 옮기고, 요소 높이의 50% 만큼 translate 하는 css 입니다.

bang9 replied:
실제로 어떻게 동작하시는지 확인하시고 적용해보시면 좋을 것 같습니다.

chohongm replied:
top: 0;
transform: translateY(50%);

둘다 껐을 때
<img width="418" alt="Screenshot 2024-07-18 at 3 02 14 PM" src="https://github.com/user-attachments/assets/021c505f-9e2d-4b7f-a939-e3278af51fae">
top: 0 만
<img width="430" alt="Screenshot 2024-07-18 at 3 02 26 PM" src="https://github.com/user-attachments/assets/09027bfa-64ee-464c-bba4-ac2148e2c839">
둘다 킬 때
<img width="436" alt="Screenshot 2024-07-18 at 3 02 31 PM" src="https://github.com/user-attachments/assets/9473d6bb-fc20-4c85-b4b6-57805dc61277">

bang9 replied:
위젯에서 말고 따로 static html 만드셔서, 컨테이너 높이가 좀 높은곳에서 테스트를 해보시죠!

요소 높이의 50% 만큼 tranform 계산을 하면 눈대중으로 대충 저 input 에서 어느쯤에 위치할지도 한번 생각해보시면 좋을 것 같습니다.

chohongm replied:
top: unset 하니 가운데 정렬해주네요. 그냥 unset 으로 다시 수정했습니다.
<img width="922" alt="Screenshot 2024-07-18 at 3 07 55 PM" src="https://github.com/user-attachments/assets/33d093b2-f606-4f97-8a1c-b64e4c41c2ca">

bang9 replied:
unset 은 위아래 패딩 값이 바뀌게 되면 깨질껄요.. 결과만 보시고 끝내지 마시고, 정확한 동작까지 이해하시는걸 추천드립니다.

chohongm replied:
이제 이해했습니다. 감사합니다. 제안주신데로 수정했습니다.
https://codesandbox.io/p/sandbox/sweet-sun-3kq7kh?file=%2Fsrc%2FApp.tsx%3A20%2C36

<img width="932" alt="Screenshot 2024-07-18 at 3 43 56 PM" src="https://github.com/user-attachments/assets/8447bca8-c88d-4f5e-9984-83867d900243">
<img width="936" alt="Screenshot 2024-07-18 at 3 44 11 PM" src="https://github.com/user-attachments/assets/646f6e4a-72b6-4c61-b1c4-aff0a136500f">
<img width="1030" alt="Screenshot 2024-07-18 at 3 42 19 PM" src="https://github.com/user-attachments/assets/40b4485e-22fa-4292-9b3a-167ac9cdb6f2">

bang9 replied:
이미 잘 알고 계실것이고, 잔소리같지만 workaround 혹은 동작 잘하니 끝! 이렇게 되면 절대 절대 안되고, 실제로 동작을 보거나 찾아보고 명확하게 동작을 이해를 하고 사용을 해야 나중에 응용해서도 사용이 가능합니다. 어떻게 동작하는가에 대한 호기심 그 자체가 엔지니어링 역량 성장에도 도움이 되고, 개인적으론 엔지니어의 필수 소양이라 생각하니 한번 마음에 새겨두시면 좋을 것 같습니다! 👍

bang9 commented on src/components/FormInput.tsx:
```diff
@@ -397,6 +411,15 @@ const FormInput = (props: InputProps) => {
+                          No Response
+                        </SubmittedPlaceholder>
+                      )}
                     </SubmittedTextInputContainer>
```
이거 컴포넌트 하나로 못합치나요..?

bang9 replied:
textarea 랑 겹치는것 같아서 합치면 좋을 것 같습니다!

chohongm replied:
수정 완료요!

bang9 replied:
수정 안된거같습니다

bang9 replied:
아래 코드를 컴포넌트화 하는걸 제안드린 리뷰입니다.
```tsx
<SubmittedTextInputContainer>
  <SubmittedText>{currentValue}</SubmittedText>
  {isOptionalEmpty
    ? <NoResponseText>No Response</NoResponseText>
    : <SubmittedText>{currentValue}</SubmittedText>
  }
  {isValid && (
    <CheckIconContainer>
      <Icon type={IconTypes.DONE} fillColor={IconColors.SECONDARY_2} width="20px" height="20px" />
    </CheckIconContainer>
  )}
</SubmittedTextInputContainer>
```

chohongm replied:
아래처럼 묶어봤는데 확인 부탁드립니다.
```
{
              const currentValue = values.length > 0 ? values[0] : '';
              const isOptionalEmpty = !required && values.length === 0;
              return (
                <InputContainer>
                  {isSubmitted ? (
                    <SubmittedTextInputContainer isTextarea={layout === 'textarea'}>
                      {isOptionalEmpty
                        ? <NoResponseText>No Response</NoResponseText>
                        : <SubmittedText>{currentValue}</SubmittedText>
                      }
                      {isValid && (
                        <CheckIconContainer>
                          <Icon type={IconTypes.DONE} fillColor={IconColors.SECONDARY_2} width="20px" height="20px" />
                        </CheckIconContainer>
                      )}
                    </SubmittedTextInputContainer>
                  ) : (
                    layout === 'textarea'
                      ? <TextArea
                        className="sendbird-input__input"
                        required={required}
                        disabled={disabled}
                        value={currentValue}
                        onFocus={handleFocus}
                        onBlur={handleBlur}
                        onChange={(event) => {
                          const value = event.target.value;
                          onChange(value ? [value] : []);
                        }}
                        placeholder={!disabled ? placeHolder : ''}
                      />
                      : <Input
                        type={layout === 'number' ? 'text' : layout}
                        className="sendbird-input__input"
                        name={name}
                        required={required}
                        disabled={disabled}
                        value={currentValue}
                        onFocus={handleFocus}
                        onBlur={handleBlur}
                        onChange={(event) => {
                          const value = event.target.value;
                          onChange(value ? [value] : []);
                        }}
                        placeholder={!disabled ? placeHolder : ''}
                      />
                  )}
                </InputContainer>
              );
            }
```

bang9 replied:
이러면 삼항연산자가 너무 많아지면서 더 복잡도가 높아지는 것 같은데요.. ㅠㅠ
처음에 말씀드린 부분만 컴포넌트로 빼는게 좋을것 같습니다

bang9 commented on src/components/FormInput.tsx:
```diff
@@ -363,6 +368,15 @@ const FormInput = (props: InputProps) => {
                           <Icon type={IconTypes.DONE} fillColor={IconColors.SECONDARY_2} width="20px" height="20px" />
                         </CheckIconContainer>
                       )}
+                      {placeHolder && !currentValue && (
```
placeholder 의 유무 체크는 왜 필요한건가요? 논의된 내용 보면, 그냥 옵셔널 필드에 응답을 제출하지 않으면 No Response 를 나타내는 변경사항처럼 보입니다.

bang9 replied:
현재는 `SubmittedText` 에 `currentValue` 를 보여주고 있는데요
옵셔널 + currentValue 가 없으면 여기에 `No response` 를 보여주고 스타일을 조정할 순 없는걸까요?

보여지는 영역은 동일한데, 따로 라벨을 추가해서 absolute 로 따로 관리를 해야하나 싶어서요.. 두개의 정확한 위치도 달라질 수 있구요

chohongm replied:
placeHolder 는 빼는게 맞습니다. 캐치 감사합니다!
굿 아이디어입니다. 그렇게 수정하겠습니다.

chohongm replied:
수정 완료요!

chohongm committed:
apply feed

bang9 commented on src/components/FormInput.tsx:
```diff
@@ -358,25 +357,21 @@ const FormInput = (props: InputProps) => {
             }
             case 'textarea': {
               const currentValue = values.length > 0 ? values[0] : '';
+              const isOptionalEmpty = !required && values.length === 0;
```
요 로직이면, currentValue 에 `''` 빈 스트링이 들어있어도 제출로 치는거죠?

bang9 replied:
요 케이스에 대해서 디자인 컨펌 한번 받으면 좋을거같네융

chohongm replied:
제출로 치는거죠를 이해못했어요. 제출된 폼은 isSubmitted 로 체크하게 되구요, currentValue 값은 ui 로 보여주기 위함이고, submit 할때는 values (default is []) 로 서브밋을 하게 됩니다.
```
onChange={(values) => {
              setFormValues(([...newInputs]) => {
                newInputs[index] = {
                  ...newInputs[index], // Create a new object for the updated item
                  draftValues: values,
                  errorMessage: (() => {
                    if (!item.isValid(values)) {
                      return stringSet.FORM_ITEM_INVALID;
                    }
                    if (required && values.length === 0) {
                      return stringSet.FORM_ITEM_REQUIRED;
                    }
                    return null;
                  })(),
                };
                return newInputs; // Return the new array
              });
            }}
```

bang9 replied:
제출된 FormItem 의 values 에 `['']` 이렇게 값이 들어있는 케이스입니다.

이 케이스에서는 values.length 가 0 이상이기 때문에 isOptionalEmpty 에 해당이 안될거고
이 경우에 `No response` 텍스트 대신 `''` 빈 텍스트가 나타나게 될거라서요

이 경우에도 빈 텍스트 대신 `No response` 를 보여주는게 UX 적으로 좋을지에 대한 이야기였습니다. (이 부분을 디자인 컨펌)

chohongm replied:
지금 코드로직이 아래처럼 구현되있기때문에
```
onChange(value ? [value] : []);
```
values 에 [''] 가 들어갈수 없습니다

bang9 replied:
여긴 제출된 form 을 보여주는곳이라서, 다른 플랫폼에서 그렇게 보낸경우 들어갈 수 있지 않나요?
다른 플랫폼과 해당 로직이 동일하면 패스해도 될거같습니다 (빈 스트링은 `['']` 가 아니라 `[]` 를 보낸다.)

bang9 replied:
아 이거 위젯이라 여기선 큰 상관 없겠네요, uikit 쪽에서는 체크되어야 할 것 같습니다.

bang9 replied:
https://github.com/sendbird/sendbird-uikit-react/pull/1168/files#r1682262138

여기로 옮겼습니다. 확인하시고, 요 대화는 Resolve 하셔도 될 것 같습니다. 레이즈 한번 부탁드립니다~

bang9 commented on src/components/FormInput.tsx:
```diff
@@ -358,25 +357,21 @@ const FormInput = (props: InputProps) => {
                   {isSubmitted ? (
                     <SubmittedTextInputContainer isTextarea={true}>
-                      <SubmittedText>{currentValue}</SubmittedText>
+                      {isOptionalEmpty
+                        ? <NoResponseText>No Response</NoResponseText>
+                        : <SubmittedText>{currentValue}</SubmittedText>
+                      }
```
NoResponseText 와 SubmittedText 를 따로 관리하지 않고, 하나로 못합치려나요?

chohongm replied:
css property 와 value 둘다 바뀌게 되어서 isOptionalEmpty 를 한번에 하기 위해서 위처럼 구현했는데, 좀 더 개선하자면 아래처럼 하는정도가 생각나네요. 더 좋은 방법이 있다면 제시 부탁드립니다.
```
const SubmittedTextComponent = ({ isOptionalEmpty, currentValue }) => {
  return (
    isOptionalEmpty
      ? <NoResponseText>No Response</NoResponseText>
      : <SubmittedText>{currentValue}</SubmittedText>
  );
}
```

bang9 replied:
SubmittedText 에서 `empty:booolean` 같은걸 props 로 받아서 색상만 처리해주는건 어떨까요?

```tsx
<SubmittedText empty={!!currentValue}>
  {currentValue ?? 'NoResponse'}
</SubmittedText/>


<SubmittedText empty={isOptionalEmpty}>
  {isOptionalEmpty ? 'NoResponse' : currentValue}
</SubmittedText/>
```

chohongm replied:
요렇게 수정됐어요
```
 <>
  <SubmittedTextInputComponent layout={layout} currentValue={currentValue} isValid={isValid} />
  {placeHolder && !currentValue && <Placeholder>No Response</Placeholder>}
</>
```
uikit처럼 Placeholder component 를 따로 만들어서 absolute 로 띄워주는 방식입니다. 요걸 안쓰고 element placeholder 를 쓰면 multi line 일때 조금 짤려보이는 이슈가 있더라구요.

bang9 replied:
placeholder && 체크는 필요한건가유?

chohongm replied:
오 없어야 합니다. 캐치 감사합니다. 수정해서 푸시했고, uikit 도 수정해서 푸시했습니다.

chohongm committed:
fix

chohongm commented on src/components/CustomChannelComponent.tsx:
```diff
@@ -292,6 +292,9 @@ export function CustomChannelComponent() {
                 />
+                {message.messageId === lastMessage?.messageId &&
+                  true &&
+                  message.data && <MessageDataContent messageData={message.data} />}
```
Fixes https://sendbird.atlassian.net/browse/AC-2782

bang9 replied:
티켓하고, 코드 한번 더 체크 부탁드립니다.

chohongm replied:
<img width="487" alt="Screenshot 2024-07-24 at 3 26 47 PM" src="https://github.com/user-attachments/assets/67b64af4-5945-4d59-9a00-af70e2be8035">
픽스 완료요!

chohongm committed:
put back

chohongm committed:
fix

chohongm committed:
fix

chohongm committed:
fix

chohongm committed:
format

chohongm committed:
fix

chohongm committed:
fix

chohongm committed:
add input mode

chohongm committed:
fix https://sendbird.atlassian.net/browse/AC-3252

chohongm committed:
Fix broken dark theme for MessageDataContent

chohongm committed:
Change submitMessageForm() to non-deprecated

chohongm committed:
fix https://sendbird.atlassian.net/browse/AC-3347

chohongm committed:
fix

chohongm committed:
Apply message form changes in uikit

chohongm committed:
format

chohongm commented:
@bang9 요거 리뷰 준비 됐으니 부탁드립니다. [sdk 배포](https://github.com/sendbird/chat-js/pull/1518)가 먼저 나가야 요거 배포 할수 있어요!

chohongm committed:
fix

chohongm commented:
runformat 은 로컬에서 고치고 푸시했는데도 실패하네요..?

bang9 commented:
> runformat 은 로컬에서 고치고 푸시했는데도 실패하네요..?

왜인지는 모르겠으나, 리암 PC 에서만 결과가 다르게 나오는것 같네요 `yarn format` 실행하신거 맞죠?

chohongm commented:
> 왜인지는 모르겠으나, 리암 PC 에서만 결과가 다르게 나오는것 같네요 `yarn format` 실행하신거 맞죠?

넵

bang9 commented:
> > 왜인지는 모르겠으나, 리암 PC 에서만 결과가 다르게 나오는것 같네요 `yarn format` 실행하신거 맞죠?
> 
> 넵

한번 해결을 시도해보시겠어요?

chohongm committed:
fix

chohongm commented:
제 랩탑은 prettier version 3.3.3을 쓰고 있는데 warning 을 캐치못하는거 같습니다. @bang9 의 prettier 버전은 몇인가요?
아니네요. npm install -g prettier@2.8.8 로 downgrade 해도 똑같습니다.

@bang9 와 싱크해서 yarn add prettier@latest 로 3.3.3 버전이 설치되니 format 이 워닝까지 잡아줘서 이렇게 푸시가 됐습니다.

chohongm committed:
format fix

bang9 commented on src/components/FormInput.tsx:
```diff
@@ -34,25 +34,26 @@ const OptionalText = styled.span`
       `solid 1px ${errorMessage ? theme.borderColor.formInput.error : theme.borderColor.formInput.default}`} !important;
-    ::placeholder {
-      color: ${({ theme }) => theme.textColor.placeholder};
-    }
```
placeholder 색상 제거는 의도된걸까요?

bang9 commented on src/components/FormInput.tsx:
```diff
@@ -64,6 +65,21 @@ interface ChipProps {
 }
 
+const Placeholder = styled.div`
+  width: calc(100% - 26px); // (12px side padding + 1px border) * 2 = 26px
+  height: calc(100% - 16px); // (7px padding + 1px border) * 2 = 16px
+  overflow: hidden;
+  text-overflow: ellipsis;
+  word-break: break-word;
+  position: absolute;
+  pointer-events: none;
+  top: 8px;
+  left: 13px;
```
흠.. 이거 나중에 너무 유지보수가 힘들어지지 않을까요..?
패딩이나 보더가 바뀌면 이걸 놓치지 않고 캐치할 수 있을지가 걱정되네요

chohongm replied:
absolute 로 띄워주려면 어쩔수 없어보입니다. 다른 방법은 placeholder property 로 가는건데 요거는 multi line 으로 placeholer string 이 보여질때 맨 밑줄이 살짝 걸쳐 보이는 이슈가 있어서 위 방식으로 수정했습니다

bang9 replied:
혹시 살짝 걸쳐보인다는게 어떻게 보이는걸까용? placeholder 의 line height 를 조정하는 방법은 안되려나요?

chohongm replied:
<img width="683" alt="Screenshot 2024-07-31 at 9 30 37 AM" src="https://github.com/user-attachments/assets/fd1879f7-79f9-4de5-9ae6-d601727ce4e7">
위에 첫번째 인풋 살짝 걸쳐서보이는거 보이시죠? 저렇게 나왔습니다. line height 도 디자인에 정해진걸 따라가려면 안될것 같아요.

chohongm replied:
또 두번째 Input 도 위 아래 패딩이 다르게 보여서 ui 상으로 이상해 보입니다

bang9 commented on src/components/MessageDataContent.tsx:
이건 따로 작업을 빼주시는게 좋지 않았을까요? form 관련된거에 섞이니까 리뷰하기가 좀 힘들어지네요 @.@

chohongm replied:
아 그렇네요 요거는 PR을 따로 만들겠습니다

bang9 commented on src/components/messages/FormMessage.tsx:
```diff
@@ -68,14 +83,10 @@ export default function FormMessage(props: Props) {
   const hasError = formValues.some(({ errorMessage }) => !!errorMessage);
+  const isButtonDisabled = ((isInvalidated || !isInputFocused) && hasError) || isSubmitted;
 
   const handleSubmit = useCallback(async () => {
```
handleSubmit 에 useCallback 제거해도 될거같습니다

bang9 commented on src/components/messages/FormMessage.tsx:
```diff
@@ -128,12 +133,18 @@ export default function FormMessage(props: Props) {
             style={style}
             placeHolder={placeholder}
             values={item.submittedValues ?? draftValues}
+            isInvalidated={isInvalidated}
```
`isInvalidated` 는 어떤 로직을 위해서 추가된걸까요?

chohongm replied:
["실시간 validation 을 Submit 이후 or 최초 validation 이후부터"](https://sendbird.slack.com/archives/C0772Q0BB8Q/p1721206742529359)

요 조건을 충족하기 위해서 추가됐습니다

bang9 commented on src/components/FormInput.tsx:
```diff
@@ -390,30 +421,29 @@ const FormInput = (props: InputProps) => {
+                          onChange(value ? [value] : []);
+                        }}
+                      />
+                      {!disabled && placeHolder && !currentValue && <Placeholder>{placeHolder}</Placeholder>}
```
disabled 면 placeholder 에 아무것도 안보여주는게 맞나여?

chohongm replied:
넵 input 이 disabled 되는 경우는 submit 이 된 경우입니다. 이때는 submitted value 나 no response 가 보여져야 합니다. 그래서 placeholder 가 있어도 보여지면 안됩니다.

bang9 replied:
그럼 `!isSubmitted` 를 쓰는게 맞지 않나유?
근데 애초에 isSubmitted 가 false 인 경우에만 이쪽이 렌더링 될거라서, 의미가 없는 조건 체크인거같아요

bang9 replied:
FormInput 이라는 컴포넌트의 비즈니스적인 역할로 생각했을때 isSubmitted 와 disabled 프롭이 모두 필요한가요?
제출 전(입력 가능), 제출 완료(입력 안됨) 두가지 상태만 있을것 같아서 질문드림다.

chohongm replied:
흠 맞는 말씀이십니다. 확인해볼게요.

chohongm replied:
disabled property 를 없앴습니다 @bang9 uikit 도 똑같이 커밋해서 푸시하겠습니다.

chohongm committed:
Fix overflow text ellipsis

chohongm committed:
Fix form input invalidation logic.

chohongm committed:
revert before merge

chohongm committed:
update typing bubble container border radius

chohongm committed:
apply feed

chohongm committed:
format

chohongm committed:
fix form disable mechanism

chohongm committed:
feat: form fallback message (#352)

## Changes
- Added fallback user message for invalid form version

ticket: https://sendbird.atlassian.net/browse/AC-3624

## Additional Notes
- 

## Checklist
Before requesting a code review, please check the following:
- [x] **[Required]** CI has passed all checks.
- [x] **[Required]** A self-review has been conducted to ensure there
are no minor mistakes.
- [x] **[Required]** Unnecessary comments/debugging code have been
removed.
- [x] **[Required]** All requirements specified in the ticket have been
accurately implemented.
- [x] Ensure the ticket has been updated with the sprint, status, and
story points.

bang9 commented on src/components/messages/FormMessage.tsx:
```diff
@@ -30,7 +34,18 @@ const Root = styled.div`
+    const disabledBgColor = theme.bgColor.formButton.disabled;
+    return {
+      'background-color': theme.bgColor.incomingMessage,
+      '.sendbird-button.sendbird-button__disabled': {
```
만약 저 classname 을 사용하지 않는 다른 button 컴포넌트를 사용하게 되면 스타일이 깨지게 되니, 이런식으로 의존하는 코드를 최대한 지양하면 좋을 것 같습니다. (classname 으로 disabled 를 판단하는 방법 말고 다른 방법을 사용하면 좋을 것 같습니다. )

chohongm replied:
!important 로 overwrite 하는 방법은 어떠신가요?
```
const SubmitButton = styled(Button)`
  width: 100%;
  ${({ theme, disabled }) => {
    if (disabled) {
      const disabledBgColor = `${theme.bgColor.formButton.disabled} !important`;
      return {
        'background-color': disabledBgColor,
        '&:hover': {
          'background-color': disabledBgColor,
        },
      };
    }
  }};
`;
```

bang9 replied:
색상이 적용이 안되어서 그런건가용? specificity 를 올려도 안되나요?

chohongm replied:
.sendbird-button__disabled 여기에 정의된 css 가 우선순위로 적용되기 때문에 위에 제가 제안드린 솔루션을 사용할 경우에 !important 를 넣지 않으면 적용이 안됩니다.
```
&.sendbird-button__disabled {
    cursor: not-allowed;
    @include themed() {
      background-color: t(bg-2);
      border: none;
    }

    &:hover {
      @include themed() {
        background-color: t(bg-2);
      }
    }
  }
```

bang9 replied:
예, 그래서 specificity 를 올려서 시도해보셨냐고 질문을 드렸습니다

chohongm replied:
음 specificity 방법에 대한 그림이 잘 안 떠올라서 좀 더 구체적으로 설명주시면 적용해보겠습니다.

bang9 replied:
예전에 관련하여 공유드렸던 링크입니다~!
- https://sendbird.slack.com/archives/C050Z6DA70A/p1714017440288599?thread_ts=1711433846.470019&cid=C050Z6DA70A

chohongm replied:
넵 specificity 가 뭔지는 이해했는데 올리는 방법이 위 경우에 정확히 어떻게 해야될지 몰라서 (!important 가 가장 쉬운 솔루션) 조금 생각해보겠습니다.

chohongm replied:
disabled 가 true 일때만 inline style 에 주입하는 방법밖에는 떠오르질 않네요. 요 방법보다 좋은 방법 아시면 공유 부탁드립니다 🙏

bang9 replied:
disabled 인 경우에만 `&&& { ...disabled styles }` 가 들어가면 될거같습니다.

chohongm replied:
앗 예전에 설명주신거에 스샷에 &&& 방법이 나왔있었네요! 제가 스샷을 제대로 안봐서 힌트를 못 봤습니다 🙏 감사합니다 저도 요거 읽어보겠습니다!
<img width="919" alt="Screenshot 2024-08-28 at 5 56 43 PM" src="https://github.com/user-attachments/assets/53fe2c04-f1e8-4335-a44a-dbb63cc17983">

chohongm replied:
윽 아래 방법이 안먹히네요..
<img width="833" alt="Screenshot 2024-08-28 at 6 08 21 PM" src="https://github.com/user-attachments/assets/b91b8b97-913a-4443-915a-c61369d669fa">

return `&&& { ... }` 도 안 먹히네요


edit: 아래가 됩니다!
```
  &&& {
    ${({ theme, disabled }) => {
      if (disabled) {
        const disabledBgColor = `${theme.bgColor.formButton.disabled};`;
        return {
          'background-color': disabledBgColor,
          '&:hover': {
            'background-color': disabledBgColor,
          },
        };
      }
    }};
  }
```

chohongm committed:
apply feeds

chohongm committed:
apply feed

bang9 commented on src/main.tsx:
Note: revert before merge

bang9 reviewed (approved):
lgtm

chohongm committed:
remove test code
